src/snapred/backend/data/Indexer.py

```diff
@@ -0,0 +1,373 @@
+from snapred.backend.dao.calibration.CalibrationRecord import CalibrationRecord
+from snapred.backend.dao.indexing.CalculationParameters import CalculationParameters
+from snapred.backend.dao.indexing.IndexEntry import IndexEntry
+from snapred.backend.dao.indexing.Record import Record
+from snapred.backend.dao.indexing.Versioning import (
+    VERSION_DEFAULT,
+    VERSION_DEFAULT_NAME,
+    VERSION_START,
+    VersionedObject,
+)
+from snapred.backend.dao.normalization.Normalization import Normalization
+from snapred.backend.dao.normalization.NormalizationRecord import NormalizationRecord
+from snapred.backend.dao.reduction.ReductionRecord import ReductionRecord
+from snapred.backend.log.logger import snapredLogger
+from snapred.meta.mantid.AllowedPeakTypes import StrEnum
+from snapred.meta.mantid.WorkspaceNameGenerator import ValueFormatter as wnvf
+from snapred.meta.redantic import parse_file_as, write_model_list_pretty, write_model_pretty
+
+logger = snapredLogger.getLogger(__name__)
+
+
+"""
+    The Indexer will automatically track versions and produce the next and current versions.
+    This is intended to have responsibility over one state/resolution/workflow combination,
+    (e.g state abcd123efg, lite mode, calibration), and for that combination will keep tabs on
+    the index and the directory tree to determine appropriate version numbers.
+
+    All saving or loading of indices, records, or calculation parameter files should be
+    handled through the appropriate Indexer.
+
+    When saving, the Indexer will always save at the version attached to the object.
+    If this version is invalid, it will throw an error and not save.
+
+    The Indexer version list will only update when both a record and a corresponding inex entry
+    have been written.
+"""
+
+
+class IndexerType(StrEnum):
+    DEFAULT = ""
+    CALIBRATION = "Calibration"
+    NORMALIZATION = "Normalization"
+    REDUCTION = "Reduction"
+
+
+# the record type for each indexer type
+RECORD_TYPE = {
+    IndexerType.CALIBRATION: CalibrationRecord,
+    IndexerType.NORMALIZATION: NormalizationRecord,
+    IndexerType.REDUCTION: ReductionRecord,
+    IndexerType.DEFAULT: Record,
+}
+
+# the params type for each indexer type
+PARAMS_TYPE = {
+    IndexerType.CALIBRATION: Calibration,
+    IndexerType.NORMALIZATION: Normalization,
+    IndexerType.REDUCTION: CalculationParameters,
+    IndexerType.DEFAULT: CalculationParameters,
+}
+
+
+class Indexer:
+    rootDirectory: Path
+    index: Dict[int, IndexEntry]
+
+    indexerType: IndexerType
+
+    ## CONSTRUCTOR / DESTRUCTOR METHODS ##
+
+    @validate_call
+    def __init__(self, *, indexerType: IndexerType, directory: Path | str) -> None:
+        self.indexerType = indexerType
+        self.rootDirectory = Path(directory)
+        self.index = self.readIndex()
+        self.dirVersions = self.readDirectoryList()
+        self.reconcileIndexToFiles()
+
+    def __del__(self):
```

Comment:
This is just a safety net right?

Reply:
Yeah, if something does wrong it will save the index before closing